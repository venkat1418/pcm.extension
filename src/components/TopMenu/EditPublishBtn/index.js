// TODO: Need to refactor to move the logic to PCM server.

import React, { Component } from 'react';
import {connect} from 'react-redux';
import {bindActionCreators} from 'redux';
import { Loader } from 'react-overlay-loader';
import {NotificationManager} from 'react-notifications';
import 'react-overlay-loader/styles.css';
import PropTypes from 'prop-types';
import {PageStatusService} from '../../../services/PageStatusService';
import {AssetsStatusService} from '../../../services/AssetsStatusService';
import {toggleEditMode} from '../../../actions';
import {AuditLogService} from '../../../services/AuditLogService';
import errHandler from '../../../helper/errHandler';
import ConfirmDialogModal from '../../modals/ConfirmDialogModal';
import messages from '../../../messages/notification';
import popupMessages from '../../../messages/popupMessages';
import Loading from '../../modals/Loading';

export class EditPublishButton extends Component {

    constructor(props){
        super(props);
        this.onUnlockPage = this.onUnlockPage.bind(this);
        this.onUnlockAssets = this.onUnlockAssets.bind(this);
        this.state={
            assetsLockedBy: {},
            showAssetsStatusModal: false,
            pending: false,
            showCommonModal: false,
            showPublishModal: false
        }

    }

    openAssetsModal = ()=>{
        this.setState({ showAssetsStatusModal: true });
    }
    closeAssetsModal = ()=>{
        this.setState({ showAssetsStatusModal: false });
    }

    edit = () => {
        document.getElementById('pcmBody').style.marginLeft = "25%";
        document.getElementById('pcmBody').style.marginRight = "15%";
        if(this.state.pending) return false;
        this.setState({ pending: true });
            // AssetsStatusService.check(this.props.currentMetadata.cmsAuthToken, this.props.currentMetadata.assetsId, this.props.userData.uid).then(resp=>{
    AssetsStatusService.check(this.props.currentMetadata.cmsAuthToken, this.props.currentMetadata.assetsId, this.props.userData,this.props.currentMetadata.pageId).then(resp=>{
            this.setState({ pending: false });
            if(resp.status !== 200){
                NotificationManager.error(`Status is not 200!\n\nCurrent Status is ${resp.status}`)
                return false;
            }
            if(resp.data.error) {
                NotificationManager.error(`Error while checking asset status!`);
                return false;
            }
            if(resp.data.user.length>0){
               // this.onUnlockPage();
               this.setState({ pending: false });
               this.props.toggleEditMode(true);
               this.closeAssetsModal();
               document.getElementById("pcmBody").style.pointerEvents = "none";
            } else if(resp.data.concurrent.length>0){
                this.setState({ assetsLockedBy: resp.data.concurrent[0].user });
                this.openAssetsModal();
            }
        })
        .catch( err=>{ this.setState({ pending: false }); errHandler.notification(err); });
        return null;
    };

    onUnlockAssets(){
        this.setState({ pending: true });
        // AssetsStatusService.unlockThenLock(this.props.currentMetadata.cmsAuthToken, this.props.currentMetadata.assetsId, this.props.userData.uid).then(resp=>{
            AssetsStatusService.unlockThenLock(this.props.currentMetadata.cmsAuthToken, this.props.currentMetadata.assetsId, this.props.userData,this.props.currentMetadata.pageId).then(resp=>{
            this.setState({ pending: false });
            if(resp.status !== 200){
                NotificationManager.error(`Status is not 200!\n\nCurrent Status is ${resp.status}`)
                return false;
            }
            if(resp.data.error){
                NotificationManager.error(`Error while assets lock & unlock!`)
                return false;
            }
            if(resp.data.user.length>0){
               // this.onUnlockPage();
               this.setState({ pending: false });
               this.props.toggleEditMode(true);
               this.closeAssetsModal();
               document.getElementById("pcmBody").style.pointerEvents = "none";
            }
            else{
                NotificationManager.error(`Unlocking asset is failed!`)
            }
        })
        .catch( err=>{ this.setState({ pending: false }); errHandler.notification(err); });
    }

    onUnlockPage(){
        this.setState({ pending: true });
        PageStatusService.unlockPage(this.props.currentMetadata.pageId, this.props.userData).then(resp=>{
            this.setState({ pending: false });
            if(resp.data === 'yes'){
                const logData={
                    action: 'unlock_page',
                    pageId: this.props.currentMetadata.pageId,
                    user: this.props.userData,
                    updatedAt: new Date().getTime()
                }
                AuditLogService.logAuditLog(logData).then(resp=>{});
                this.setState({ pending: true });
                PageStatusService.lockPage(this.props.currentMetadata.pageId, this.props.userData).then(resp=>{
                    this.setState({ pending: false });
                    if(resp.data === 'yes'){
                        const logData={
                            action: 'lock_page',
                            pageId: this.props.currentMetadata.pageId,
                            user: this.props.userData,
                            updatedAt: new Date().getTime()
                        }
                        AuditLogService.logAuditLog(logData).then(resp=>{});
                        this.props.toggleEditMode(true);
                        this.closeAssetsModal();
                        document.getElementById("pcmBody").style.pointerEvents = "none";
                    }
                })
                .catch(err=>{ this.setState({ pending: false }); NotificationManager.error(`Error while locking page!`) })
            }
        })
        .catch(err=>{ this.setState({ pending: false }); NotificationManager.error(`Error while unlocking page!`) })
    }

    publish = () => {
        this.setState({ showPublishModal: true });
    };

    onPublish = () =>{
      if(this.state.pending) return false;
      this.setState({ pending: true, showPublishModal: false });
      Promise.all([
          PageStatusService.unlockPage(this.props.currentMetadata.pageId, this.props.userData),
          AssetsStatusService.removeLock(this.props.currentMetadata.cmsAuthToken, this.props.currentMetadata.assetsId, this.props.userData.uid)
      ]).then(resp=>{
          this.setState({ pending: false });
          if(resp[0].data !== 'yes'){
              NotificationManager.error(`Unlocking page is failed!`)
              return false;
          }
          if(resp[1].status !== 200){
              NotificationManager.error(`Status is not 200!\n\nCurrent Status is ${resp[1].status}!`)
              return false;
          }
          if(!resp[1].data){
              NotificationManager.error(`Unlocking asset is failed!`)
              return false;
          }
          const logData={
              action: 'unlock_page',
              pageId: this.props.currentMetadata.pageId,
              user: this.props.userData,
              updatedAt: new Date().getTime()
          }
          AuditLogService.logAuditLog(logData).then(resp=>{});
          this.props.toggleEditMode(false);
          NotificationManager.success(messages.publishMessage);
          document.getElementById("pcmBody").style.pointerEvents = "auto";
      })
      .catch(err=>{ this.setState({ pending: false }); })
    }

    onCancelPublish = () =>{
        this.setState({showPublishModal: false});
    }

    discard = (e)=>{
        if(this.state.pending) return false;
        if(this.props.isPageDirty) {
            e.preventDefault();
            this.setState({showCommonModal: true});
        } else {
            this.props.toggleEditMode(false);
            document.getElementById('pcmBody').style.marginLeft = '20%';
            document.getElementById('pcmBody').style.marginRight = '20%';
            NotificationManager.info(messages.discardButtonMsg_unlock);
        }
        return null;
    }

    onDiscard = (e) =>{
        this.setState({showCommonModal: false});
        this.props.toggleEditMode(false);
        NotificationManager.info(messages.discardButtonMsg_confirm);
        document.getElementById('pcmBody').style.marginLeft = '20%';
        document.getElementById('pcmBody').style.marginRight = '20%';
    }
    onCancelDiscard = () =>{
        this.setState({showCommonModal: false});
    }

    render() {
      const lockedBy = `${this.state.assetsLockedBy.name} is currently editing this page`;
        return (
          <div className="editPublishComp">
            {
                    /* <button className="edit_toggle_button" disabled={!this.props.isLogin} onClick={(this.props.isEditMode? this.discard:this.edit)}>{(this.props.isEditMode?"DISCARD":"EDIT")}</button> */

                    this.props.isEditMode?
                      <button className="edit_toggle_button" disabled={!this.props.isLogin} onClick={this.discard}>DISCARD</button>
                    :
                      <button className="edit_toggle_button" disabled={!this.props.isLogin} onClick={this.edit}>EDIT</button>

                }
            <button className="publishButton" disabled={!this.props.isEditMode || this.props.noOfOpenStories !== 0} onClick={this.publish}>PUBLISH</button>
            {/*
                    this.props.isEditMode?
                    <img alt="" className={"lock "+(this.props.isLogin ? "": "disabledItem")} height="40px" src="images/lock.png" onClick={this.props.isLogin? this.discard : ''} />
                    :
                    <img alt="" className={"lock "+(this.props.isLogin ? "": "disabledItem")} height="40px" src="images/unlock2.png" onClick={this.props.isLogin? this.edit : ''}/>
                */}
            <ConfirmDialogModal
              showModal={this.state.showAssetsStatusModal}
              mainDescription={popupMessages.editLocked.mainDescription}
              originalHeadline={lockedBy}
              onConfirm={this.onUnlockAssets}
              onCancel={this.closeAssetsModal}
              confirmLabel='UNLOCK'
                />
            <ConfirmDialogModal
              showModal={this.state.showCommonModal}
              mainDescription={popupMessages.discardPage.mainDescription}
              originalHeadline={popupMessages.discardPage.originalHeadline}
              onConfirm={this.onDiscard}
              onCancel={this.onCancelDiscard}
              confirmLabel='DISCARD'
                />
            <ConfirmDialogModal
              showModal={this.state.showPublishModal}
              mainDescription={popupMessages.publishPage.mainDescription}
              originalHeadline={popupMessages.publishPage.originalHeadline}
              onConfirm={this.onPublish}
              onCancel={this.onCancelPublish}
              confirmLabel='PUBLISH'
                />

                <Loading
                    showLoadingModal={this.state.pending}
                  />
          </div>
        );
    }
};

const mapStateToProps = (state)=>({
        isLogin: state.isLogin,
        isEditMode: state.isEditMode,
        userData: state.userData,
        currentMetadata: state.currentMetadata,
        isPageDirty: state.data.isPageDirty
    })
const mapDispatchToProps = (dispatch)=>bindActionCreators({
        toggleEditMode
    }, dispatch)

EditPublishButton.propTypes = {
  userData: PropTypes.object.isRequired,
  currentMetadata: PropTypes.object.isRequired,
  toggleEditMode: PropTypes.func.isRequired,
  isEditMode: PropTypes.bool.isRequired,
  isLogin: PropTypes.bool.isRequired,
  noOfOpenStories: PropTypes.number,
  isPageDirty: PropTypes.bool.isRequired
}
export default connect(mapStateToProps, mapDispatchToProps)(EditPublishButton);
