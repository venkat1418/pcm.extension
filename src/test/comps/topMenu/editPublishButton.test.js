
import { shallow } from 'enzyme';
import React from 'react';
import { NotificationManager } from 'react-notifications';
import {AssetsStatusService} from '../../../services/AssetsStatusService';
import {PageStatusService} from '../../../services/PageStatusService';
import {AuditLogService} from '../../../services/AuditLogService';
import EditPublishBtn from "../../../components/TopMenu/EditPublishBtn";

beforeAll(() => {
  NotificationManager.error = jest.fn();
  NotificationManager.info = jest.fn();
  NotificationManager.success = jest.fn();


});

test('Test the renders of EditPublishBtn', () => {

   let isLogin = false ;
   let isEditMode = false;


   let wrapper = shallow(<EditPublishBtn isLogin={isLogin} isEditMode={isEditMode}/>);
   //Create a snapshot with the default render of the component
   expect(wrapper).toMatchSnapshot();

   // check the button rendered , they will be in disable state however when isLogin = false
   expect(wrapper.find('button').at(0).text()).toBe('EDIT');
   expect(wrapper.find('button').at(1).text()).toBe('PUBLISH');

   expect(wrapper.find('button').at(0).prop('disabled')).toBe(true);
   expect(wrapper.find('button').at(1).prop('disabled')).toBe(true);

   isLogin = true ;
   isEditMode = false ;

   wrapper = shallow(<EditPublishBtn isLogin={isLogin} isEditMode={isEditMode}/>);

   expect(wrapper.find('button').at(0).text()).toBe('EDIT');
   expect(wrapper.find('button').at(1).text()).toBe('PUBLISH');

   expect(wrapper.find('button').at(0).prop('disabled')).toBe(false);
   expect(wrapper.find('button').at(1).prop('disabled')).toBe(true);


   isLogin = true ;
   isEditMode = true ;

   wrapper = shallow(< EditPublishBtn isLogin={isLogin} isEditMode={isEditMode}/>);

   expect(wrapper.find('button').at(0).text()).toBe('DISCARD');
   expect(wrapper.find('button').at(1).text()).toBe('PUBLISH');

   expect(wrapper.find('button').at(0).prop('disabled')).toBe(false);
   expect(wrapper.find('button').at(1).prop('disabled')).toBe(true);

   // no test few methods
   wrapper.instance().openAssetsModal();
   expect(wrapper.state().showAssetsStatusModal).toBe(true);

   wrapper.instance().closeAssetsModal();
   expect(wrapper.state().showAssetsStatusModal).toBe(false);

});

test('Test the edit() callback of EditPublishBtn with non 200 status', done => {
  expect.assertions(1);
  const props = {
     toggleLogin: jest.fn(),
     toggleEditMode: jest.fn(),
     currentMetadata : {
       assetsId : 99
     },
     userData : {
       uid : 99
     }
  }
  AssetsStatusService.check = jest.fn();
  AssetsStatusService.check
 .mockReturnValueOnce(new Promise(function(resolve, reject){
         resolve({status: 502});
   }))

 let wrapper = shallow(< EditPublishBtn {...props}/>);
  wrapper.instance().edit();

   setTimeout(() => {
       expect(NotificationManager.error).toHaveBeenCalledWith("Status is not 200!\n\nCurrent Status is 502")
       done();
     }, 10);
});

test('Test the edit() callback of EditPublishBtn with error in resp', done => {
  expect.assertions(1);
  const props = {
     toggleLogin: jest.fn(),
     toggleEditMode: jest.fn(),
     currentMetadata : {
       assetsId : 99
     },
     userData : {
       uid : 99
     }
  }
  AssetsStatusService.check = jest.fn();
  AssetsStatusService.check
   .mockReturnValueOnce(new Promise(function(resolve, reject){
        resolve({status: 200,
                  data: {
                    error: "Failed!!"
                  } });
  }))

 let wrapper = shallow(< EditPublishBtn {...props}/>);

  wrapper.instance().edit();
  setTimeout(() => {
       expect(NotificationManager.error).toHaveBeenCalledWith("Error while checking asset status!")
       done();
  }, 10);

});

test('Test the edit() callback of EditPublishBtn , successfully entered edit mode', done => {
  expect.assertions(1);
  const props = {
     toggleLogin: jest.fn(),
     toggleEditMode: jest.fn(),
     currentMetadata : {
       assetsId : 99
     },
     userData : {
       uid : 99
     }
  }

  var dummyElement = document.createElement('pcmBody');
  var dummyElement2 = document.createElement('top-menu');
  document.getElementById = jest.fn();
  document.getElementById
      .mockReturnValueOnce(dummyElement)
      .mockReturnValueOnce(dummyElement2)


  AssetsStatusService.check = jest.fn();
  AssetsStatusService.check
  .mockReturnValueOnce(new Promise(function(resolve, reject){
       resolve({status: 200,
                 data: {
                   user: [{uid:99}],
                 }});
   }))


 let wrapper = shallow(< EditPublishBtn {...props}/>);

 wrapper.instance().onUnlockPage = jest.fn();

  wrapper.instance().edit();
  setTimeout(() => {
      // expect(NotificationManager.success).toHaveBeenCalledWith("Successfully entered to edit mode")
      expect(wrapper.instance().onUnlockPage.mock.calls.length).toBe(1);
       done();
  }, 10);

});

test('Test the edit() callback of EditPublishBtn, enter locking ', done => {
  expect.assertions(2);
  const props = {
     toggleLogin: jest.fn(),
     toggleEditMode: jest.fn(),
     currentMetadata : {
       assetsId : 99
     },
     userData : {
       uid : 99
     }
  }


  var dummyElement = document.createElement('pcmBody');
  var dummyElement2 = document.createElement('top-menu');
  document.getElementById = jest.fn();
  document.getElementById
      .mockReturnValueOnce(dummyElement)
      .mockReturnValueOnce(dummyElement2)

  AssetsStatusService.check = jest.fn();
  AssetsStatusService.check
   .mockReturnValueOnce(new Promise(function(resolve, reject){
        resolve({status: 200,
                  data: {
                    user: [],
                    concurrent : [{user : [{uid:99}]}]
                  }});
    }))

 let wrapper = shallow(< EditPublishBtn {...props}/>);
  wrapper.instance().edit();
  setTimeout(() => {
       expect(wrapper.state().showAssetsStatusModal).toBe(true);
       expect(wrapper.state().assetsLockedBy).toEqual([ { uid: 99 } ]);
       done();
  }, 10);


});


test('Test the publish() callback of EditPublishBtn, with unlocking page failed', done => {
      const props = {
                 currentMetadata : {
                   assetsId : 99,
                   pageId : 99
                 },
                 userData : {
                   uid : 99
                 }
              }

      AssetsStatusService.removeLock = jest.fn();
      AssetsStatusService.removeLock
       .mockReturnValueOnce(new Promise(function(resolve, reject){
            resolve({});
        }))

      PageStatusService.unlockPage = jest.fn();
      PageStatusService.unlockPage
      .mockReturnValueOnce(new Promise(function(resolve, reject){
            resolve({data: "no"});
        }))


     let wrapper = shallow(< EditPublishBtn {...props} />);
     wrapper.instance().publish();

     expect.assertions(1);

          setTimeout(() => {
              expect(NotificationManager.error).toHaveBeenCalledWith("Unlocking page is failed!")
               done();
          }, 10);

});

test('Test the publish() callback of EditPublishBtn, with status not 200', done => {
   const props = {
        currentMetadata : {
          assetsId : 99,
          pageId : 99
        },
        userData : {
          uid : 99
        }
    }

  AssetsStatusService.removeLock = jest.fn();
  AssetsStatusService.removeLock
    .mockReturnValueOnce(new Promise(function(resolve, reject){
        resolve({status: 502});
     }))

  PageStatusService.unlockPage = jest.fn();
  PageStatusService.unlockPage
    .mockReturnValueOnce(new Promise(function(resolve, reject){
        resolve({data: "yes"});
     }))


  let wrapper = shallow(< EditPublishBtn {...props} />);
  wrapper.instance().publish();

  expect.assertions(1);

  setTimeout(() => {
      expect(NotificationManager.error).toHaveBeenCalledWith("Status is not 200!\n\nCurrent Status is 502!")
        done();
  }, 10);

});

test('Test the publish() callback of EditPublishBtn, no data', done => {
       const props = {
            currentMetadata : {
              assetsId : 99,
              pageId : 99
            },
            userData : {
              uid : 99
            }
        }

      AssetsStatusService.removeLock = jest.fn();
      AssetsStatusService.removeLock
        .mockReturnValueOnce(new Promise(function(resolve, reject){
            resolve({status: 200});
         }))

      PageStatusService.unlockPage = jest.fn();
      PageStatusService.unlockPage
        .mockReturnValueOnce(new Promise(function(resolve, reject){
            resolve({data: "yes"});
         }))


      let wrapper = shallow(< EditPublishBtn {...props} />);
      wrapper.instance().publish();

      expect.assertions(1);

      setTimeout(() => {
          expect(NotificationManager.error).toHaveBeenCalledWith("Unlocking asset is failed!")
            done();
      }, 10);

});

test('Test the publish() callback of EditPublishBtn, publish sucessfully', done => {

   AuditLogService.logAuditLog = jest.fn();

   AuditLogService.logAuditLog
      .mockReturnValue(new Promise(function(resolve, reject){
           resolve({resp: ''});
       }));

    const props = {
        currentMetadata : {
          assetsId : 99,
          pageId : 99
        },
        userData : {
          uid : 99
        },
        toggleEditMode: jest.fn(),
    }


  var dummyElement = document.createElement('pcmBody');
  document.getElementById = jest.fn();
  document.getElementById
      .mockReturnValueOnce(dummyElement)

    AssetsStatusService.removeLock = jest.fn();
    AssetsStatusService.removeLock
      .mockReturnValueOnce(new Promise(function(resolve, reject){
         resolve({status: 200,data: {
             user: [{uid:99}],
         }});
      }))

    PageStatusService.unlockPage = jest.fn();
    PageStatusService.unlockPage
        .mockReturnValueOnce(new Promise(function(resolve, reject){
            resolve({data: "yes"});
         }))


    let wrapper = shallow(< EditPublishBtn {...props} />);
     wrapper.instance().publish();

     expect.assertions(2);

    setTimeout(() => {
        expect(NotificationManager.success).toHaveBeenCalledWith("Published successfully!")
        expect(props.toggleEditMode).toHaveBeenCalledWith(false);
        done();
        }, 10);

});

test('Test the onUnlockPage', done => {
  AuditLogService.logAuditLog = jest.fn();

  AuditLogService.logAuditLog
     .mockReturnValue(new Promise(function(resolve, reject){
          resolve({resp: ''});
      }));
  const props = {
          currentMetadata : {
            assetsId : 99,
            pageId : 99
          },
          userData : {
            uid : 99
          },
          toggleEditMode: jest.fn(),
      }

  PageStatusService.unlockPage = jest.fn();
  PageStatusService.lockPage = jest.fn();

  PageStatusService.unlockPage.mockReturnValueOnce(new Promise(function(resolve, reject){
     resolve({data: 'yes'});
  }))

  PageStatusService.lockPage.mockReturnValueOnce(new Promise(function(resolve, reject){
     resolve({data: 'yes'});
  }))
  let wrapper = shallow(< EditPublishBtn {...props} />);
  wrapper.instance().onUnlockPage();

  expect.assertions(2);

 setTimeout(() => {
     expect(props.toggleEditMode).toHaveBeenCalledWith(true);
     expect(wrapper.state().showAssetsStatusModal).toBe(false);
     done();
     }, 10);
})

test('Test the onUnlockPage, catch the unlocking error ', done => {
  AuditLogService.logAuditLog = jest.fn();

  AuditLogService.logAuditLog
     .mockReturnValue(new Promise(function(resolve, reject){
          resolve({resp: ''});
      }));
  const props = {
          currentMetadata : {
            assetsId : 99,
            pageId : 99
          },
          userData : {
            uid : 99
          },
      }

  PageStatusService.unlockPage = jest.fn();
  PageStatusService.lockPage = jest.fn();

  PageStatusService.unlockPage.mockReturnValueOnce(new Promise(function(resolve, reject){
     reject({});
  }))


  let wrapper = shallow(< EditPublishBtn {...props} />);
  wrapper.instance().onUnlockPage();

  expect.assertions(1);

 setTimeout(() => {
     expect(NotificationManager.error).toHaveBeenCalledWith("Error while unlocking page!")
     done();
     }, 10);
})

test('Test the onUnlockPage, catch the locking error ', done => {
  AuditLogService.logAuditLog = jest.fn();

  AuditLogService.logAuditLog
     .mockReturnValue(new Promise(function(resolve, reject){
          resolve({resp: ''});
      }));
  const props = {
          currentMetadata : {
            assetsId : 99,
            pageId : 99
          },
          userData : {
            uid : 99
          },
      }

  PageStatusService.unlockPage = jest.fn();
  PageStatusService.lockPage = jest.fn();

  PageStatusService.unlockPage.mockReturnValueOnce(new Promise(function(resolve, reject){
     resolve({data: 'yes'});
  }))

  PageStatusService.lockPage.mockReturnValueOnce(new Promise(function(resolve, reject){
     reject({});
  }))


  let wrapper = shallow(< EditPublishBtn {...props} />);
  wrapper.instance().onUnlockPage();

  expect.assertions(1);

 setTimeout(() => {
     expect(NotificationManager.error).toHaveBeenCalledWith("Error while locking page!")
     done();
     }, 10);
})

test('Test the edit, catch the Authentication error ', done => {
  const props = {
     toggleLogin: jest.fn(),
     toggleEditMode: jest.fn(),
     currentMetadata : {
       assetsId : 99
     },
     userData : {
       uid : 99
     }
  }
  AssetsStatusService.check = jest.fn();
  AssetsStatusService.check
 .mockReturnValueOnce(new Promise(function(resolve, reject){
         reject({response: {status : 403}});
   }))

 let wrapper = shallow(< EditPublishBtn {...props}/>);
  wrapper.instance().edit();

   setTimeout(() => {
       expect(NotificationManager.error).toHaveBeenCalledWith("Authentication Error! 403")
       done();
     }, 10);

})

test('Test the edit, catch the CMS server error ', done => {
  const props = {
     toggleLogin: jest.fn(),
     toggleEditMode: jest.fn(),
     currentMetadata : {
       assetsId : 99
     },
     userData : {
       uid : 99
     }
  }
  AssetsStatusService.check = jest.fn();
  AssetsStatusService.check
 .mockReturnValueOnce(new Promise(function(resolve, reject){
         reject({response: {status : 503}});
   }))

 let wrapper = shallow(< EditPublishBtn {...props}/>);
  wrapper.instance().edit();

   setTimeout(() => {
       expect(NotificationManager.error).toHaveBeenCalledWith("CMS Server Error! 503")
       done();
     }, 10);

})

test('Test the onUnlockAssets() callback of EditPublishBtn with non 200 status', done => {
  expect.assertions(1);
  const props = {
     currentMetadata : {
       assetsId : 99,
       cmsAuthToken:99
     },
     userData : {
       uid : 99
     },
  }
  AssetsStatusService.unlockThenLock = jest.fn();
  AssetsStatusService.unlockThenLock
 .mockReturnValueOnce(new Promise(function(resolve, reject){
         resolve({status: 502});
   }))

 let wrapper = shallow(< EditPublishBtn {...props}/>);
  wrapper.instance().onUnlockAssets();

   setTimeout(() => {
       expect(NotificationManager.error).toHaveBeenCalledWith("Status is not 200!\n\nCurrent Status is 502")
       done();
     }, 10);
})


test('Test the onUnlockAssets() callback of EditPublishBtn with error in resp', done => {
  expect.assertions(1);
  const props = {
     currentMetadata : {
       assetsId : 99,
       cmsAuthToken : 99
     },
     userData : {
       uid : 99
     }
  }
  AssetsStatusService.unlockThenLock = jest.fn();
  AssetsStatusService.unlockThenLock
   .mockReturnValueOnce(new Promise(function(resolve, reject){
        resolve({status: 200,
                  data: {
                    error: "Failed!!"
                  } });
  }))

 let wrapper = shallow(< EditPublishBtn {...props}/>);

  wrapper.instance().onUnlockAssets();
  setTimeout(() => {
       expect(NotificationManager.error).toHaveBeenCalledWith("Error while assets lock & unlock!")
       done();
  }, 10);

})

test('Test the onUnlockAssets() callback of EditPublishBtn , successfully entered edit mode', done => {
  expect.assertions(1);
  const props = {
     currentMetadata : {
       assetsId : 99,
       cmsAuthToken : 99
     },
     userData : {
       uid : 99
     }
  }

  AssetsStatusService.unlockThenLock = jest.fn();
  AssetsStatusService.unlockThenLock
  .mockReturnValueOnce(new Promise(function(resolve, reject){
       resolve({status: 200,
                 data: {
                   user: [{uid:99}],
                 }});
   }))


 let wrapper = shallow(< EditPublishBtn {...props}/>);

 wrapper.instance().onUnlockPage = jest.fn();

  wrapper.instance().onUnlockAssets();
  setTimeout(() => {
      // expect(NotificationManager.success).toHaveBeenCalledWith("Successfully entered to edit mode")
      expect(wrapper.instance().onUnlockPage.mock.calls.length).toBe(1);
       done();
  }, 10);

})

test('Test the onUnlockAssets() callback of EditPublishBtn , unlocking asset failed', done => {
  expect.assertions(1);
  const props = {
     currentMetadata : {
       assetsId : 99,
       cmsAuthToken : 99
     },
     userData : {
       uid : 99
     }
  }

  AssetsStatusService.unlockThenLock = jest.fn();
  AssetsStatusService.unlockThenLock
  .mockReturnValueOnce(new Promise(function(resolve, reject){
       resolve({status: 200,
                 data: {
                   user: [],
                 }});
   }))


 let wrapper = shallow(< EditPublishBtn {...props}/>);

 wrapper.instance().onUnlockPage = jest.fn();

  wrapper.instance().onUnlockAssets();
  setTimeout(() => {
       expect(NotificationManager.error).toHaveBeenCalledWith("Unlocking asset is failed!")
       done();
  }, 10);

})

test('Test the onUnlockAssets(), catch the Authentication error ', done => {
  const props = {
     currentMetadata : {
       assetsId : 99
     },
     userData : {
       uid : 99
     }
  }
  AssetsStatusService.unlockThenLock = jest.fn();
  AssetsStatusService.unlockThenLock
 .mockReturnValueOnce(new Promise(function(resolve, reject){
         reject({response: {status : 403}});
   }))

 let wrapper = shallow(< EditPublishBtn {...props}/>);
  wrapper.instance().onUnlockAssets();

   setTimeout(() => {
       expect(NotificationManager.error).toHaveBeenCalledWith("Authentication Error! 403")
       done();
     }, 10);

})

test('Test the onUnlockAssets, catch the CMS server error ', done => {
  const props = {
     currentMetadata : {
       assetsId : 99
     },
     userData : {
       uid : 99
     }
  }
  AssetsStatusService.unlockThenLock = jest.fn();
  AssetsStatusService.unlockThenLock
 .mockReturnValueOnce(new Promise(function(resolve, reject){
         reject({response: {status : 503}});
   }))

 let wrapper = shallow(< EditPublishBtn {...props}/>);
  wrapper.instance().onUnlockAssets();

   setTimeout(() => {
       expect(NotificationManager.error).toHaveBeenCalledWith("CMS Server Error! 503")
       done();
     }, 10);

})


test('Test the discard() callback of EditPublishBtn', () => {
  let props = {
    toggleEditMode: jest.fn()
}

  let wrapper = shallow(< EditPublishBtn {...props}/>);
  wrapper.instance().discard();
  expect(props.toggleEditMode).toHaveBeenCalledWith(false);
  expect(NotificationManager.info).toHaveBeenCalledWith('Successfully entered view mode and page unlocked!')

  props = {
    toggleEditMode: jest.fn(),
    isPageDirty : true
  }
  const event = { preventDefault: () => {} };

  wrapper = shallow(< EditPublishBtn {...props}/>);
  wrapper.instance().discard(event);
});
