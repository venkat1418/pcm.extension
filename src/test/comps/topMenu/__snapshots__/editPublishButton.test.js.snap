// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test the renders of EditPublishButton 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <EditPublishButton
    isEditMode={false}
    isLogin={false}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <button
          className="edit_toggle_button"
          disabled={true}
          onClick={[Function]}
        >
          EDIT
        </button>,
        <button
          className="publishButton"
          disabled={true}
          onClick={[Function]}
        >
          PUBLISH
        </button>,
        <AssetsStatusModal
          lockedBy={Object {}}
          onCancel={[Function]}
          onConfirm={[Function]}
          showModal={false}
        />,
        <div />,
      ],
      "className": "editPublishComp",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "EDIT",
          "className": "edit_toggle_button",
          "disabled": true,
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "EDIT",
        "type": "button",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "PUBLISH",
          "className": "publishButton",
          "disabled": true,
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "PUBLISH",
        "type": "button",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "lockedBy": Object {},
          "onCancel": [Function],
          "onConfirm": [Function],
          "showModal": false,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <button
            className="edit_toggle_button"
            disabled={true}
            onClick={[Function]}
          >
            EDIT
          </button>,
          <button
            className="publishButton"
            disabled={true}
            onClick={[Function]}
          >
            PUBLISH
          </button>,
          <AssetsStatusModal
            lockedBy={Object {}}
            onCancel={[Function]}
            onConfirm={[Function]}
            showModal={false}
          />,
          <div />,
        ],
        "className": "editPublishComp",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "EDIT",
            "className": "edit_toggle_button",
            "disabled": true,
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "EDIT",
          "type": "button",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "PUBLISH",
            "className": "publishButton",
            "disabled": true,
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "PUBLISH",
          "type": "button",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "lockedBy": Object {},
            "onCancel": [Function],
            "onConfirm": [Function],
            "showModal": false,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
